task01-anton11nahhas-main/tasks_week1_Anton_Nahhas/src/Vector.cpp

```diff
@@ -0,0 +1,272 @@
+	if (this != &other)// if lhs == rhs return this object.
+	{
+		if (m_size != other.m_size)//if lhs size != rhs size make new array with new size.
+		{
+			m_size = other.m_size;
+			int* newArray = new int[m_size];
+			delete[] m_array;
+			m_array = newArray;
+		}
+		for (int i = 0; i < m_size; i++)
+			m_array[i] = other.m_array[i];//copy rhs to lhs
+	}
+	return *this;
+}
+
+//--------------------------------------------------------------------------------
+/*move assignment operator, receives a temporary rvalue other and moves it to lvalue*/
+Vector& Vector::operator=(Vector&& other)noexcept{
+	/*std::cout << "move operator" << std::endl;*/
+	if (this != &other)// if lhs == rhs return this object.
+	{
+		delete[] m_array;
+		if (m_size != other.m_size)//if lhs size != rhs size make new array with new size.
+		{	
+			m_size = other.m_size;
+			m_capacity = other.m_size;
+		}
+		m_array = other.m_array;
+		other.m_size = 0;
+		other.m_capacity = 0;
+		other.m_array = nullptr;
+	}
+	return *this;
+}
+
+//--------------------------------------------------------------------------------
+/*this function receives a value adds a cell at the end and pushes the value there,
+the function does not return*/
+void Vector::push_back(const int& value) {
+	
+	std::size_t tempSize = m_size +  1;
+	if (tempSize >= m_capacity ) {
+		resize(tempSize, value);
+	}
+
+	m_array[m_size - 1] = value;
+
+}
+
+//--------------------------------------------------------------------------------
+/*this function resizes the vector to the desired count, if count is bigger than curr
+size then resize to new size and fill with 0's if the vouns is smaller than the curr
+size then only show the new size values, the function does not return */
+void Vector::resize(std::size_t count) {
+	resize(count, 0);
+	
+}
+
+//--------------------------------------------------------------------------------
+/*this function resizes the vector to the desired count and value, if the count is bigger
+than the current size then change to new size and fill with the value provided, if the
+count is smaller than the current size then only show the first count elements*/
+void Vector::resize(std::size_t count, const int& value) {
+	std::size_t tempSize = m_size;
+
+	if (m_capacity < count) {
+		while (m_capacity < count) {
+			if (m_capacity <= 128) {
+				if (m_capacity == 0) { m_capacity = 2; }
+				else {
+					m_capacity *= 2;
+				}
+			}
+			else { m_capacity *= 1.5; }
+		}
+	}
+	if (m_size > count) {
+		m_size = count;
+		int* newArray = new int[m_size];
+		for (int i = 0; i < m_size; i++) {
+			newArray[i] = m_array[i];
+		}
+		delete[] m_array;
+		m_array = std::move(newArray);
+	}
+	else {
+		m_size = count;
+		int* newArray = new int[m_size];
+		for (std::size_t i = 0; i < tempSize; i++) {
+			newArray[i] = m_array[i];
+		}
+		for (std::size_t i = tempSize; i < count; i++) {
+			newArray[i] = value;
+		}
+		delete[] m_array;
+		m_array = std::move(newArray);
+	}
+}
```

Comment:
In the second if statement, the loop copies the first count elements of the array into a new array, but it would be simpler and more efficient to use std::copy from the <algorithm> header.